presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.metadata.Signature;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+import com.facebook.presto.sql.planner.plan.WindowNode;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkState;
+
+public class MergeIdenticalWindows
+        implements PlanOptimizer
+{
+    public PlanNode optimize(PlanNode plan,
+            Session session,
+            Map<Symbol, Type> types,
+            SymbolAllocator symbolAllocator,
+            PlanNodeIdAllocator idAllocator)
+    {
+        return SimplePlanRewriter.rewriteWith(new Rewriter(), plan, new Context());
+    }
+
+    private static class Context
+    {
+        private Map<WindowNode.Specification, ImmutableMap.Builder<Symbol, FunctionCall>> functions =
+                new HashMap<>();
+        private Map<WindowNode.Specification, ImmutableMap.Builder<Symbol, Signature>> signatures =
+                new HashMap<>();
+
+        private boolean collectFunctions(WindowNode window)
+        {
+            WindowNode.Specification key = window.getSpecification();
+
+            // Either both maps must contain the specification, or neither.
+            checkState(functions.containsKey(key) == signatures.containsKey(key),
```

Comment:
Couldn't we merge both maps to avoid such a consistency checks?
E.g. Map<Specification, ImmutableMap.Builder<Symbol, Pair<FunctionCall, Signature>>>` or, instead of `Pair` you can use some specialized, more descriptive structure?
You even have it already - `FunctionPair`.